### PR TITLE
Use frictionless framework v5

### PR DESCRIPTION
**Pull request by aivuk**
Update the extension to use [Frictionless Framework](https://github.com/frictionlessdata/framework) version 5. Change the validate function from Goodtables to the Frictionless equivalent.

TODO: 

* Update the Report component to display the validation results using the [new React component](https://components.frictionlessdata.io/?path=/story/components-report--invalid).
* Add more documentation in how to pass the option parameters to the [`validate`](https://framework.frictionlessdata.io/docs/guides/validating-data.html) function. With goodtables you could pass skip_rows and headers options that are not supported by the current `validate`.
* Add tests for the new validate options

### CODE REVIEW

**Comment by amercader** [comment]
@aivuk this PR should be against `dev-v2` right?

**Comment by aivuk** [comment]
> @aivuk this PR should be against `dev-v2` right?

yes, sorry about that. I already changed it to dev-v2 now

File: ckanext/validation/jobs.py [hunk 1]
```diff
@@ -117,18 +125,23 @@ def run_validation_job(resource):
          'validation_timestamp': validation.finished.isoformat()})
 
 
-def _validate_table(source, _format=u'csv', schema=None, **options):
+def _validate_table(source, _format='csv', schema=None, **options):
+
+    # TODO: Search if there is an equivalent way to use a proxy in Frictionless Framework v5
```
Comment (ThrawnCA) [inline ckanext/validation/jobs.py:130]: I'd like this to be addressed before merging the change; it has security implications.

Comment (aivuk) [reply ckanext/validation/jobs.py]: Fixed handling the http_proxy correctly with frictionless framework v5 in [28608e9](https://github.com/frictionlessdata/ckanext-validation/pull/69/commits/28608e9deead5d8604874ba7db35eb8acb496bd9)

**Review by ThrawnCA — APPROVED** [review]

File: ckanext/validation/jobs.py (hunk 1 above)
Comment (ThrawnCA) [reply ckanext/validation/jobs.py]: 👍

File: ckanext/validation/jobs.py [hunk 2]
```diff
@@ -54,54 +55,62 @@ def run_validation_job(resource):
...
-    if schema and isinstance(schema, str):
-        if schema.startswith('http'):
-            r = requests.get(schema)
-            schema = r.json()
+        source = resource['url']
+
+    schema = resource.get('schema')
+    if schema:
+        if isinstance(schema, str):
+            if schema.startswith('http'):
```
Comment (ThrawnCA) [inline ckanext/validation/jobs.py:83]: So, if `schema` is a string that doesn't start with "http" then it's preserved unchanged?

File: README.md [hunk 3]
```diff
@@ -51,7 +51,7 @@ If you are eager to get started, jump to the [Installation](#installation) and [
 
 ## Versions supported and requirements
 
-This extension is currently tested in CKAN 2.8 and CKAN 2.9.
+This extension is currently tested in CKAN 2.9.
```
Comment (amercader) [inline README.md:54]:
```suggestion
This extension is currently tested in CKAN 2.9 (Python 3).
```

File: ckanext/validation/jobs.py [hunk 4]
```diff
@@ -7,7 +7,8 @@
 
 import requests
 from sqlalchemy.orm.exc import NoResultFound
-from goodtables import validate
+# from goodtables import validate
+from frictionless import validate, system, Report
```
Comment (amercader) [inline ckanext/validation/jobs.py:11]:
```suggestion
from frictionless import validate, system, Report
```

File: ckanext/validation/jobs.py [hunk 5]
```diff
@@ -54,54 +55,62 @@ def run_validation_job(resource):
...
+
+    schema = resource.get('schema')
+    if schema:
+        if isinstance(schema, str):
+            if schema.startswith('http'):
+                r = requests.get(schema)
+                schema = r.json()
         else:
-            schema = json.loads(schema)
+            schema = json.dumps(schema)
```
Comment (amercader) [inline ckanext/validation/jobs.py:87]: This logic is wrong, `schema` can be:
* A python dict, passed directly to the action (eg tests) 
* A string, in that case it can either be:
   * A URL (we need to get the contents and load the JSON)
   * Something else: we assume is a JSON string, so we need to load it (should be valid because it has passed the validators)
We should check all these cases and end up with a schema dict. I assume `validate()` expects a dict for the schema, so I'm not sure why you are using `dumps()` here

File: ckanext/validation/jobs.py [hunk 6]
```diff
@@ -54,54 +55,62 @@ def run_validation_job(resource):
...
-        validation.status = u'error'
-        validation.error = {
-            'message': '\n'.join(report['warnings']) or u'No tables found'}
+        validation.status = 'error'
+        validation.report = json.dumps(report)
+        if 'tables' in report: 
+            validation.error = {
+                'message': [str(err) for err in report['tables'][0]['errors']] if len(report['tables'][0]['errors']) > 0 else 'No tables found'}
+        else:
+            validation.error = {'message': []}
```
Comment (amercader) [inline ckanext/validation/jobs.py:113]:
```suggestion
            validation.error = {'message': ['Error running validation']}
```

Is there another key in the report that can give a more meaningful message?

File: ckanext/validation/jobs.py [hunk 7]
```diff
@@ -117,18 +126,21 @@ def run_validation_job(resource):
...
     http_session = options.pop('http_session', None) or requests.Session()
 
     use_proxy = 'ckan.download_proxy' in t.config
     if use_proxy:
         proxy = t.config.get('ckan.download_proxy')
-        log.debug(u'Download resource for validation via proxy: %s', proxy)
+        log.debug('Download resource for validation via proxy: %s', proxy)
         http_session.proxies.update({'http': proxy, 'https': proxy})
-    report = validate(source, format=_format, schema=schema, http_session=http_session, **options)
+        frictionless_context['http_session'] = http_session
```
Comment (amercader) [inline ckanext/validation/jobs.py:139]: The `http_session` can be also used without `use_proxy` being True, so this needs to be one less level of indentation

File: ckanext/validation/jobs.py [hunk 8]
```diff
@@ -117,18 +126,21 @@ def run_validation_job(resource):
          'validation_timestamp': validation.finished.isoformat()})
 
 
-def _validate_table(source, _format=u'csv', schema=None, **options):
+def _validate_table(source, _format='csv', schema=None, **options):
 
+    frictionless_context = { 'trusted': True }
```
Comment (amercader) [inline ckanext/validation/jobs.py:131]: I could not find any documentation in the fd v5 docs for what this property does, can you explain it?

File: ckanext/validation/logic.py [hunk 9]
```diff
@@ -656,26 +656,31 @@ def _run_sync_validation(resource_id, local_upload=False, new_resource=True):
         {u'ignore_auth': True},
         {u'resource_id': resource_id})
 
-    report = validation['report']
+    if validation['report']:
```
Comment (amercader) [inline ckanext/validation/logic.py:659]: Is there an instance when `validation["report"]` is not set? This is our own ckanext-validation object so even if frictionless doesn't return an object for some reason we could store a custom one like `{"valid": False, "errors": [...]}` or something similar. IIUC in the goodtables-based code we always had a `report` key

File: ckanext/validation/templates/scheming/form_snippets/resource_schema.html [hunk 10]
```diff
@@ -1,7 +1,6 @@
 {% import 'macros/form.html' as form %}
 
   {% set value = data[field.field_name] %}
-  {% set is_url = value and value[4:]|lower == 'http' %}
```
Comment (amercader) [inline ckanext/validation/templates/scheming/form_snippets/resource_schema.html:4]: This is still used below right?

File: ckanext/validation/tests/test_form.py [hunk 11]
```diff
@@ -405,8 +405,8 @@ def test_resource_form_create_invalid(self, app):
         )
 
         assert "validation" in response.body
-        assert "missing-value" in response.body
-        assert "Row 2 has a missing value in column 4" in response.body
+        assert "missing-cell" in response.body
+        assert "This row has less values compared to the header row" in response.body
```
Comment (amercader) [inline ckanext/validation/tests/test_form.py:409]: Is there somewhere in the new error messages that says what row and column are the failing ones (ie 2 and 4)?

File: ckanext/validation/tests/test_jobs.py [hunk 12]
```diff
@@ -140,8 +140,9 @@ def test_job_run_invalid_stores_validation_object(self, mock_validate):
             .one()
         )
 
-        assert validation.status == "failure"
-        assert validation.report == INVALID_REPORT
+        assert validation.status == "error"
+        print(json.dumps(json.loads(validation.report), indent=4))
```
Comment (amercader) [inline ckanext/validation/tests/test_jobs.py:144]:
```suggestion
```

File: ckanext/validation/tests/test_jobs.py [hunk 13]
```diff
@@ -201,103 +202,110 @@ def test_job_run_valid_stores_status_in_resource(self, mock_validate):
...
-
-        validation = (
-            Session.query(Validation)
-            .filter(Validation.resource_id == resource["id"])
-            .one()
-        )
-
-        assert validation.report["valid"] is True
+# TODO: The vollowing test is not valid on frictionless v5. 
+# The local path is not hidden and there is no warning for more
```
Comment (amercader) [inline ckanext/validation/tests/test_jobs.py:206]: Do you mean than in v5 there's no need to hide the paths?

**Review by amercader — COMMENTED** [review]
Nice work @aivuk, I'll do some more testing locally but I left a first round of comments

File: ckanext/validation/jobs.py (hunk 5 above)
Comment (ThrawnCA) [reply ckanext/validation/jobs.py:86]: It looks to me like this code is handling all of those cases, except that it will not consistently result in a dict.

- Passing in a dict will follow the `else` branch and go through `json.dumps`, producing a string.
- Passing in a URL will follow both `if` branches, producing a JSON object.
- Passing a JSON string will follow the first `if`, but not the second, retaining the original string.

**Comment by amercader** [comment]
@aivuk The `Aborted (core dumped)` errors were not a Docker issue, but a max recursion error introduced by a change in CKAN core, tests are good now!

File: ckanext/validation/tests/test_jobs.py [hunk 14]
```diff
@@ -201,103 +201,110 @@ def test_job_run_valid_stores_status_in_resource(self, mock_validate):
...
-            run_validation_job(resource)
-
-        validation = (
-            Session.query(Validation)
-            .filter(Validation.resource_id == resource["id"])
-            .one()
-        )
-
-        assert validation.report["valid"] is True
+# TODO: The vollowing test is not valid on frictionless v5.
```
Comment (ThrawnCA) [inline ckanext/validation/tests/test_jobs.py:204]: Misspelled "following"

File: ckanext/validation/jobs.py (hunk 8 above)
Comment (aivuk) [reply ckanext/validation/jobs.py:141]: This is necessary when you pass a full path to the validate function, otherwise it fails. I did need a hard time to discover that because it isn't in the docs. This option is there to follow the standard https://specs.frictionlessdata.io/data-resource/#data-location that discourages the use of absolute paths

File: requirements.txt [hunk 15]
```diff
@@ -1,3 +1,5 @@
 ckantoolkit>=0.0.3
-goodtables==1.5.1
+frictionless==5.0.0b9
```
Comment (ThrawnCA) [inline requirements.txt:2]: Can we use a release version instead of beta?

**Review by ThrawnCA — CHANGES_REQUESTED** [review]

**Comment by aivuk** [comment]
@roll  In the past we could pass to to the goodtables validation function an options json object where it was possible to define which tests to skip and pass parameters about the layout of the data, as number of rows, headers, etc https://github.com/frictionlessdata/ckanext-validation/#formats-to-validate 
In this PR I'm using directly the Frictionless validate function at https://github.com/frictionlessdata/ckanext-validation/blob/ae2087656712b3198fca66c45e2f45fc2a3c1cac/ckanext/validation/jobs.py#L154 and passing a "options" json object stored in the database by the user, but I couldn't find a way to pass the parameters to customize the validations checks as https://framework.frictionlessdata.io/docs/checks/table.html. What would you suggest?

 @amercader

**Comment by roll** [comment]
Hi @aivuk,

In v5 the ad-hoc validation function got standardized:

```
schema = Schema() # create schema according to Table Schmea spec
checklist = Checklist() # create checklist according to https://framework.frictionlessdata.io/docs/framework/checklist.html (see API reference)
resource = Resource(..., schema=schema, checklist=checklist) # create resource according to Data Resource spec
```

You can consider storing the whole validation information as a resource descriptor in ckan database:

```
resource
  path
  scheme
  format
  ...
  schema
  checklist
```

and then just use `Resource(descritor).validate()`

**Comment by roll** [comment]
So in the Framework now the whole information of **data + schema + validation rules** might be stored/shared as a resource descriptor. As one chunk of metadata. I think it's quite similar to greatexpections approach =)

**Comment by roll** [comment]
UPD.
I missed Tabular Dialect in the first place. Now it's added as it's needed to manage heaer rows etc

**Comment by roll** [comment]
Also, as a part of Frictionless Application (components@2) I'm finishing visual components to create customizable editors for these metadata as:
- Resource Editor
- Dialect Editor
- Schema Editor
- Checklist Editor
So this metadata boundaries now is reflected on UI level too

**Comment by roll** [comment]
So when you need to UI part you just need to create a bootstrap modal with:

components2

which will edit the resource descriptor.

**Comment by aivuk** [comment]
> Also, as a part of Frictionless Application (components@2) I'm finishing visual components to create customizable editors for this metadata as:
> 
>     * Resource Editor
> 
>     * Dialect Editor
> 
>     * Schema Editor
> 
>     * Checklist Editor
> 
> 
> So these metadata boundaries now is reflected on UI level too

That will be great because the next steps will be to improve the UI to customize the validation steps in CKAN Resource page. I asked myself with it would be possible to use just some of the components from the Frictionless Application, like we are doing with the Report component. Thanks!

**Comment by roll** [comment]
Yea that's the plan -- fully-reusable component for all metadata Frictionless has a package/resource/schema/etc that can be mixed in different configurations. We will maintain them as a part of the application codebase but release them as a separate NPM package.

**Comment by aivuk** [comment]
@roll Is there an example of a Resource descriptor that includes the schema, checklist and dialect? Currently in ckanext-validation we have two fields saved in the database, the schema and the options, that I think that one it will not be necessary anymore. What you suggest is to just have a field where the user store all the resource descriptor information? What is the best way to create a descriptor, create the checklist separately, the layout, then the resource and save its descriptor? I want to create some descriptors with different parameters for the tests

**Comment by roll** [comment]
Note that v5 has a clear boundary between descriptors and object model so the most explicit way of working with metadata will be:

```
resource = Resource.from_descriptor()
...
descriptor = resource.to_descriptor()
```

File: ckanext/validation/tests/test_form.py (hunk 11 above)
Comment (aivuk) [reply ckanext/validation/tests/test_form.py:410]: I changed the tested with the row and field information. The format is different from the goodtables and it has the character '"' escaped inside the html.

**Comment by amercader** [comment]
@roll @aivuk Note that the current version of ckanext-validation stores two separate fields:

* `schema`: a Table Schema descriptor
* `validation_options`: what is says on the tin, were passed directly to goodtables

I think that in terms of interoperability it is important to keep the schema (which is based on a standard) separate from the validation options (which are specific to frictionless framework). The idea is that `schema` can be used by other integrations like the DataStore etc. so I'm keen on keeping it separate. We can definitely store the whole validation descriptor (schema + validation rules) in the `validation_options` and keep in sync the `schema`  subset, or we can keep them separate and combine them when we need to call the validation, which might be probably easier in terms of keeping the current fields, validators etc (for the CKAN metadata, not for the data)

Another question, is there a way of translating goodtables validation optins to framework validation descriptors? I'm thinking in terms of migrating existing installations of ckanext-validation that are using the old options (I'm assuming there are no changes at the `schema` level as it's based on Table Schema)

**Comment by aivuk** [comment]
> We can definitely store the whole validation descriptor (schema + validation rules) in the `validation_options` and keep in sync the `schema` subset, or we can keep them separate and combine them when we need to call the validation, which might be probably easier in terms of keeping the current fields, validators etc (for the CKAN metadata, not for the data)

I was thinking a bit about it and arrived at the same conclusion, I think the schema must be separate from the validation options, but I still haven't explored too much in how this split can be done and merged easily before running the validation. I will try generate some different descriptors using the framework and do some tests in how this can be done, but I think it will be straightforward.

> Another question, is there a way of translating goodtables validation optins to framework validation descriptors? I'm thinking in terms of migrating existing installations of ckanext-validation that are using the old options (I'm assuming there are no changes at the `schema` level as it's based on Table Schema)

I think we would need to write new code to do that. The problem is that I couldn't find a good documentation in the goodtables format, only these pages:

* https://github.com/frictionlessdata/goodtables.io/wiki/.goodtables.yml-reference
* https://github.com/frictionlessdata/goodtables.io/wiki/Configuring

**Comment by roll** [comment]
@amercader 
What I suggest is to store a resource desciptor according to Data Resource Standard. It has standard properties including `resource.dialect` (CSV/Table Dialect standard) and `resource.schema` (Table Schema standard). The only framework specific (custom) property will be `resource.checklist` (aka validation options)

Of course, if you think it's better to separate schema it will also work as it's also an interchangeable standard

**Comment by roll** [comment]
@aivuk 
https://github.com/frictionlessdata/framework/tree/v2

**Comment by amercader** [comment]
Thanks both for the input.
I still think keeping just the Table Schema separate makes more sense to keep the whole implementation as is. So @aivuk let's keep the separate `schema` and `validation_options` field so everything keeps working as now and piece together the resource descriptor needed by framework just before validation.
IIUC the new validation options format can be something like `{"dialect": {}, "checklist": {}"}`, which will be combined with the `schema` to create the resource descriptor needed

**Comment by roll** [comment]
I agree it's of course the quickest way. In the end, it doesn't matter how an exact implementation stores Frictionless concepts internally.

In the future, I think it still will be good to rebase on more standards / framework@5 compatible terminology (e.g. dialect is a resource/file's property, not a validation option, and also you might need to store encoding/format/compression/etc making `validation_options` basically a resource descriptor de-factor).

@amercader 
I have a more high-level question. Have you ever thought about adopting (Tabular) Data Resource Standard to be an internal part of CKAN for resource/file metadata description? We started a similar discussion with Dan Feder from DCAN and, currently, they are working on supporting something like this. In Frictionless Application, any resource is also described by Data Resource descriptor.

**Comment by amercader** [comment]
@roll see https://github.com/frictionlessdata/ckanext-validation/discussions/72

**Comment by aivuk** [comment]
@amercader @roll I did update in my last commit the plugin to parse the validation options as: 

```
{"dialect": { ...}, "checks": {...}}
```
I used "checks" instead of "checkslist" to follow the frictionless framework nomenclature. Options as "skip_errors" or "pick_errors" can also be passed. @roll do we have a documentation for how the Dialect descriptor must be written? I'm generating it using the framework as:

```
dialect = Dialect(header=True, header_rows=[1], comment_char='#')
dialect.add_control(formats.CsvControl(delimiter=';'))
dialect.to_descriptor()
```
Do we have the documentation for how to pass the checks parameters descriptors? I could find only an example here https://framework.frictionlessdata.io/docs/checks/table.html#using-declaratively. I think it would be a good idea to create a list in the framework documentation with all the possible checks and their parameters, and this can be generated automatically.

Currently in the CKAN Resource Validation Options form we don't give much information to the user in how to write the options:
Screenshot from 2022-10-24 23-40-11
We just point to the github repository of ckanext-validation. I think it would be better to add a short text, maybe a button that would open a small dialog window with some help text and point the user to a better page, probably a link to the Frictionless framework documentation. Same thing for the Data Schema field.

**Comment by roll** [comment]
Hi @aivuk!

Can you please elaborate on the first question (how to pass)?

Regarding the checks docs it's for example here - https://framework.frictionlessdata.io/docs/checks/table.html#table-dimensions - and others in the menu under "Validation Checks". All of them have a "API Reference" section. Also, @shashigharti is currently working on adding more docs to checks/steps instead of "NOTE: add docs"

**Comment by roll** [comment]
TBH I don't fully understand a new approach. If there is still a field for CKAN end-users called "validation_options" - how they will be filling it if `framework@5` doesn't even have this concept of validation options? I thought it was for internal representation, not user-facing.

**Comment by aivuk** [comment]
> TBH I don't fully understand a new approach. If there is still a field for CKAN end-users called "validation_options" - how they will be filling it if `framework@5` doesn't even have this concept of validation options? I thought it was for internal representation, not user-facing.

The ckanext-validation is running the function `validate`  from frictionless framework. If an user wants to pass parameters to the `validate` as skip_errors or pick_errors, or customize the dialect, or pass parameters to the validation checks, s/he is going to do it passing a json in the "validation options" field.

> Can you please elaborate on the first question (how to pass)?

I mean a simple list with all the checks options names as 'table-dimensions' and their parameters. The only list that I could find was for the baseline checks but there is no documentation for what each one does https://framework.frictionlessdata.io/docs/checks/baseline.html

**Comment by aivuk** [comment]
@roll Just an example, currently the user can put in the validation options field the json:

```
{
  "checks: [
    {"type": "table-dimensions", "minRows": 3},
    {"type": 'ascii-value'}
  ],
  "dialect": {
     "header": true,
     "headerRows": [1],
     "commentChar": "#",
     "csv": {
       "delimiter": ";"
     }
}

```

File: README.md [hunk 16]
```diff
@@ -125,31 +145,55 @@ To prevent the extension from adding the validation badges next to the resources
...
+fields when creating or updating datasets. The form will return an error, for
+instance, if a field is missing or it doesn't have the expected format.
+
+Data validation follows the same principle, but against the actual data
+published in CKAN, that is the contents of tabular files (Excel, CSV, etc)
+hosted in CKAN itself or elsewhere. Whenever a resource of the appropriate
+format is created or updated, the extension will validate the data against a
+collection of checks. This validation is powered by
+[Goodtables](https://github.com/frictionlessdata/goodtables-py), a very
+powerful data validation library developed by [Open Knowledge International](https://okfn.org)
```
Comment (amercader) [inline README.md:158]:
```suggestion
[Frictionless Framework](https://github.com/frictionlessdata/frictionless-py), a very
powerful data validation library developed by the [Open Knowledge Foundation](https://okfn.org) 
```

File: README.md [hunk 17]
```diff
@@ -125,31 +145,55 @@ To prevent the extension from adding the validation badges next to the resources
...
+
+Data validation follows the same principle, but against the actual data
+published in CKAN, that is the contents of tabular files (Excel, CSV, etc)
+hosted in CKAN itself or elsewhere. Whenever a resource of the appropriate
+format is created or updated, the extension will validate the data against a
+collection of checks. This validation is powered by
+[Goodtables](https://github.com/frictionlessdata/goodtables-py), a very
+powerful data validation library developed by [Open Knowledge International](https://okfn.org) 
+as part of the [Frictionless Data](https://frictionlessdata.io) project.
+Goodtables provides an extensive suite of [checks](https://github.com/frictionlessdata/goodtables-py#checks)
```
Comment (amercader) [inline README.md:160]:
```suggestion
Frictionless Framework provides an extensive suite of [checks](https://framework.frictionlessdata.io/docs/checks/baseline.html)
```
Please update if there's a better link. It would be better to link to the parent "Validation checks" header but I don't think that's possible

File: README.md [hunk 18]
```diff
@@ -125,31 +145,55 @@ To prevent the extension from adding the validation badges next to the resources
...
+that cover common issues with tabular data files.
+
+These checks include structural problems like missing headers or values, blank
+rows, etc., but also can validate the data contents themselves (see 
+[Data Schemas](#data-schemas)) or even run [custom checks](https://github.com/frictionlessdata/goodtables-py#custom-constraint).
```
Comment (amercader) [inline README.md:165]: Is this "custom checks" functionality available on v5? If not remove it from here.

File: README.md [hunk 19]
```diff
@@ -212,14 +259,24 @@ If we store this schema against a resource, it will be used to perform a more th
...
 
 
 ### Validation Options
 
-As we saw before, the validation process involves many different checks and it's very likely that what "valid" data actually means will vary across CKAN instances or datasets. The validation process can be tweaked by passing any of the [supported options](https://github.com/frictionlessdata/goodtables-py#validatesource-options) to Goodtables. These can be used to add or remove specific checks, control limits, etc.
+As we saw before, the validation process involves many different checks and
+it's very likely that what "valid" data actually means will vary across CKAN
+instances or datasets. The validation process can be tweaked by passing any of
+the [supported
+options](https://github.com/frictionlessdata/goodtables-py#validatesource-options)
```
Comment (amercader) [inline README.md:273]: Update link, or document what's possible

File: README.md [hunk 20]
```diff
@@ -509,7 +618,10 @@ def resource_validation_run_batch(context, data_dict):
 
 ### Starting the validation process manually
 
-You can start (asynchronous) validation jobs from the command line using the `paster validation run` command. If no parameters are provided it will start a validation job for all resources in the site of suitable format (ie `ckanext.validation.formats`):
+You can start (asynchronous) validation jobs from the command line using the
+`paster validation run` command. If no parameters are provided it will start a
```
Comment (amercader) [inline README.md:622]: Looks like we forgot to migrate this command and the `report `and `report-full` ones to Click :scream: 
We can do it in a separate PR

File: ckanext/validation/jobs.py (hunk 8 above)
Comment (amercader) [reply ckanext/validation/jobs.py:141]: Cool, let's add a comment then to clarify this
```suggestion
    # This option is needed to allow Frictionless Framework to validate absolute paths
    frictionless_context = { 'trusted': True }
```

File: ckanext/validation/templates/scheming/form_snippets/resource_schema.html (hunk 10 above)
Comment (amercader) [reply ckanext/validation/templates/scheming/form_snippets/resource_schema.html:4]: ping, I think this line should stay

File: ckanext/validation/tests/test_jobs.py (hunk 13 above)
Comment (amercader) [reply ckanext/validation/tests/test_jobs.py:206]: I think these tests should be reactivated no?

File: requirements.txt (hunk 15 above)
Comment (amercader) [reply requirements.txt:2]: Sure, as soon as frictionless v5 goes out of beta we'll target the stable version

File: README.md [hunk 21]
```diff
@@ -238,66 +295,107 @@ The following validation options would make validation pass:
...
+    "skip_errors": ["blank-row"]
+    "dialect":  {
+      "header": True,
+      "headerRows": [2],
+      "commentChar": "#",
+      "csv": {
+        "delimiter": ";"
+      }
+    }
 }
```
Comment (amercader) [inline README.md:307]: Can you update the example to show some `checks` as well?

**Review by amercader — COMMENTED** [review]
@aivuk this looks good, just some polishing of the docs and tests and we can merge this

We can iron out things in separate PRs

**Comment by amercader** [comment]
@roll 

> TBH I don't fully understand a new approach. If there is still a field for CKAN end-users called "validation_options" - how they will be filling it if framework@5 doesn't even have this concept of validation options? I thought it was for internal representation, not user-facing.

We just need a way for users for inputing parameters for the `validate()` call. Right now we have this existing field which accepts a JSON object so it's easy to dump everything there rather than create separate `checks`, `dialect`, `skip_errors` etc fields in the form. The end goal should be to have a proper UI for tweaking all these settings, which I assume would be another frictionless component. When that's ready we'll refactor the form to use this and internally we'll settle on a way to pass the parameters to frictionless' `validate()` function

File: README.md (hunk 17 above)
Comment (aivuk) [reply README.md:160]: Yes, it's not possible to link to the top "Validation checks", I think this is the best link for now.

File: README.md (hunk 19 above)
Comment (aivuk) [reply README.md]: For now I think the best page is https://framework.frictionlessdata.io/docs/guides/validating-data.html

File: README.md (hunk 21 above)
Comment (aivuk) [reply README.md:307]: Just did it!

**Comment by roll** [comment]
@amercader 
That totally sounds good as an intermediate step. Just wanted to ensure that we're on the same page regarding the final vision (a UI component field instead of the `validation_options` field)

File: README.md [hunk 22]
```diff
@@ -238,66 +294,111 @@ The following validation options would make validation pass:
...
+    "skip_errors": ["blank-row"]
+    "dialect":  {
+      "header": True,
+      "headerRows": [2],
+      "commentChar": "#",
+      "csv": {
+        "delimiter": ";"
+      }
+    },
+    "checks: [
```
Comment (ThrawnCA) [inline README.md:306]: Missing closing quote.

**Review by ThrawnCA — CHANGES_REQUESTED** [review]

File: .github/workflows/test.yml [hunk 23]
```diff
@@ -49,6 +49,8 @@ jobs:
       run: |
         pip install -r dev-requirements.txt
         pip install -r requirements.txt
+        pip install --no-warn-conflicts jinja2==2.10.1
```
[the remark this answers is not in the document]
Comment (amercader) [reply .github/workflows/test.yml:52]: @ThrawnCA this will be solved with upcoming requirement changes in frictionless-py. We just added this to not block the tests on this branch.

@aivuk would be good to sort this out on frictionless-py. IIRC it's just loosening up the requirements

**Comment by amercader** [comment]
Great work @aivuk ! :rocket:
